scheduler/src/cook/kubernetes/controller.clj

```diff
@@ -160,41 +189,113 @@
-                                 [:missing :pod/failed] (remove-finalization-if-set-and-delete api-client expected-state-dict pod)
+                                 ; There was a race and it completed normally before being it was killed.
+                                 [:expected/killed :pod/succeeded] (pod-has-just-completed kcc existing-state-dict)
+                                 [:expected/killed :pod/failed] (pod-has-just-completed kcc existing-state-dict)
+                                 [:expected/killed :missing] (do ; Weird. We always update datomic first. Could happen if someone manually removed stuff from kubernetes.
+                                                               (log-weird-state expected-state-dict existing-state-dict)
+                                                               (pod-was-killed kcc pod-name))
+
+                                 ; These cases are weird.
+                                 [:expected/completed :pod/waiting] (kill-task-weird api-client expected-state-dict existing-state-dict)
+                                 [:expected/completed :pod/running] (kill-task-weird api-client expected-state-dict existing-state-dict)
+
+                                 ; This is interesting. This indicates that something deleted it behind our back!
+                                 ; Other 3 cases of [{completed,starting,killed}, :missing] already handled.
+                                 [:expected/running :missing] (do
+                                                                (log/error "Something deleted" pod-name "behind our back")
+                                                                (pod-has-just-completed kcc existing-state-dict)) ; TODO: Should mark mea culpa retry
+
+                                 ; This block of :missing,* cases can only occur in testing when you're e.g., blowing away the database. 
+                                 ; The next two will go through :missing,running or :missing,waiting to :missing,failed and then be deleted.
+                                 ; TODO: These may be evidence of a bug where we process pod changes when we're starting up.
+                                 [:missing :pod/running] (kill-task-weird api-client nil pod)
+                                 [:missing :pod/waiting] (kill-task-weird api-client nil pod)
+
+                                 ; We shouldn't hit these unless we get a database rollback.
+                                 [:missing :pod/succeeded] (kill-task-weird api-client nil pod)
+                                 [:missing :pod/failed] (kill-task-weird api-client nil pod)
+
+                                 [:missing :pod/unknown] (kill-task-weird api-client expected-state-dict existing-state-dict)
+                                 [:expected/starting :pod/unknown] ; TODO: Should mark mea culpa retry
+                                 (kill-task-weird api-client (pod-has-just-completed kcc existing-state-dict) pod)
```

Comment:
We should extract this into a common function. All of these weird states appear to be doing the same thing.

Reply:
> Making one function to do these 4 tasks doesn't save any code, and would be semantically more complicated.

Calling 
`(process-weird-task api-client compute-cluster existing-state-dict pod)`
instead of 
`(kill-task-weird api-client (pod-has-just-completed compute-cluster existing-state-dict) pod)`
is comfortably less verbose and easier to read. That too at **four** call-sites! When they do different things, we can have the four different implementations and inline them accordingly.

Reply:
I'd consider if it we have a better name than 'process'. THat term is meaningless.

This is actually doing double-duty. Its writing to datomic (via pod-has-completed), and also writing to kubernetes (via kill-task). It is the only case where we do both.

Reply:
Feel free to choose a more appropriate name, I used `process-*` as an example. 3rd party resolution item :)

Reply:
Call it mark-pod-completed-and-kill-task.